pillarbox-player/src/main/java/ch/srgssr/pillarbox/player/extension/Tracks.kt

```diff
@@ -4,55 +4,64 @@
 
+import android.annotation.SuppressLint
 import androidx.media3.common.C
 import androidx.media3.common.C.TrackType
+import androidx.media3.common.Format
 import androidx.media3.common.TrackGroup
 import androidx.media3.common.Tracks
 
 /**
  * Text tracks
  */
 val Tracks.text: List<Tracks.Group>
-    get() = filterByTrackType(C.TRACK_TYPE_TEXT).mapNotNull { it.filterForced() }
+    get() = filterByTrackType(C.TRACK_TYPE_TEXT).mapNotNull { it.filterForcedAndUnsupported() }
 
 /**
  * Audio tracks.
  */
 val Tracks.audio: List<Tracks.Group>
-    get() = filterByTrackType(C.TRACK_TYPE_AUDIO)
+    get() = filterByTrackType(C.TRACK_TYPE_AUDIO).mapNotNull { it.filterUnsupported() }
 
 /**
  * Video tracks.
  */
 val Tracks.video: List<Tracks.Group>
-    get() = filterByTrackType(C.TRACK_TYPE_VIDEO)
+    get() = filterByTrackType(C.TRACK_TYPE_VIDEO).mapNotNull { it.filterUnsupported() }
 
 private fun Tracks.filterByTrackType(trackType: @TrackType Int): List<Tracks.Group> {
     return groups.filter { it.type == trackType }
 }
 
+private fun Tracks.Group.filterForcedAndUnsupported(): Tracks.Group? {
+    return filterBy { group, i -> group.getTrackFormat(i).isForced() || !group.isTrackSupported(i) }
+}
+
+internal fun Tracks.Group.filterUnsupported(): Tracks.Group? {
+    return filterBy { group, i -> !group.isTrackSupported(i) }
+}
+
+@SuppressLint("WrongConstant")
 @Suppress("SpreadOperator", "ReturnCount")
-private fun Tracks.Group.filterForced(): Tracks.Group? {
-    if (type != C.TRACK_TYPE_TEXT) return this
-    if (length == 1 && getTrackFormat(0).isForced()) return null
-    val listIndexNotForced = ArrayList<Int>(length)
+internal fun Tracks.Group.filterBy(filter: (Tracks.Group, Int) -> Boolean): Tracks.Group? {
+    if (length == 1 && filter(this, 0)) return null
+    val listIndexToKeep = ArrayList<Int>(length)
     for (i in 0 until length) {
-        val track = getTrackFormat(i)
-        if (!track.isForced()) {
-            listIndexNotForced.add(i)
+        if (!filter(this, i)) {
+            listIndexToKeep.add(i)
         }
     }
-    if (listIndexNotForced.size == length) return this
-    val count = listIndexNotForced.size
-    val formats = Array(count) {
-        getTrackFormat(listIndexNotForced[it])
-    }
-    val trackSupport = Array(count) {
-        getTrackSupport(listIndexNotForced[it])
-    }
-    val trackEnabled = Array(count) {
-        isTrackSelected(listIndexNotForced[it])
+    if (listIndexToKeep.isEmpty()) return null
+    if (listIndexToKeep.size == length) return this
+    val count = listIndexToKeep.size
+    val formats = ArrayList<Format>(count)
+    val trackSupport = IntArray(count)
+    val trackSelect = BooleanArray(count)
+    for (i in 0 until count) {
+        val trackIndex = listIndexToKeep[i]
+        formats.add(getTrackFormat(trackIndex))
+        trackSupport[i] = getTrackSupport(trackIndex)
+        trackSelect[i] = isTrackSelected(trackIndex)
     }
```

Comment:
Starts to feel like code smell a little bit, don't you think? Not sure if this can be written in a better way, though.